### PR TITLE
Provisioning of things

### PR DESCRIPTION
**Pull request by rewritten**

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -42,4 +42,13 @@ Actualizar rubygems e instalar bundler
     gem install bundler
     
 
-    
+Como instalar docker y containers en mac (WIP):
+===============================================
+
+1. instalar ansible
```
Comment (enricostano) [inline CONTRIBUTING.md]: add steps for Mac, I'll do that for Linux

File: provisioning/development-osx.yml [hunk 2]
```diff
@@ -0,0 +1,7 @@
+---
+- hosts: localhost
+  # remote_user: ubuntu
+  # sudo: yes
```
Comment (enricostano) [inline provisioning/development-osx.yml]: this is the new `become: true`, we need to understand if needed for more tasks or we should only use it for specific tasks

File: provisioning/roles/container.osx/tasks/homebrew.yml [hunk 3]
```diff
@@ -0,0 +1,16 @@
+---
+# Thanks to http://marvelley.com/blog/2014/04/11/local-provisioning-with-ansible/
+- name: Check Homebrew is installed
+  stat: path=/usr/local/bin/brew
+  register: brew_installed
+
+- name: Install Homebrew
+  shell: ruby -e "$(curl -fsSL https://raw.github.com/Homebrew/homebrew/go/install)"
```
Comment (enricostano) [inline provisioning/roles/container.osx/tasks/homebrew.yml]: is this the only way to do it? `ruby`? `which ruby`? :joy_cat:

File: provisioning/roles/container.osx/tasks/main.yml [hunk 4]
```diff
@@ -0,0 +1,3 @@
+---
+- include: homebrew.yml
+- include: xhyve.yml
```
Comment (enricostano) [inline provisioning/roles/container.osx/tasks/main.yml]: install ruby too, maybe?

**Comment by enricostano** [comment]
Reaaaalllyy niiice @rewritten :clap: :clap: :clap: 

An easy to setup development environment is a MUST to welcome contributors :muscle: 

Really loving this :heart_eyes:

**Comment by d2bit** [comment]
I use docker-compose to set up my dev environment. I didn't know about ansible+docker. Interesting!!!

File: provisioning/development-osx.yml (hunk 2 above)
Comment (rewritten) [inline provisioning/development-osx.yml]: Just a comment, it needs to go away (the `become: yes` bit is only in the tasks that need it).

File: provisioning/roles/container.osx/tasks/homebrew.yml (hunk 3 above)
Comment (rewritten) [inline provisioning/roles/container.osx/tasks/homebrew.yml]: Blindly pasted from that link. Suggestions welcome.

File: provisioning/roles/container.osx/tasks/main.yml (hunk 4 above)
Comment (rewritten) [inline provisioning/roles/container.osx/tasks/main.yml]: Right.

File: provisioning/roles/container.osx/tasks/xhyve.yml [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    dest: /usr/local/bin/docker-machine-driver-xhyve
+    checksum: "sha1:9af9650a697336f1b00ef64475d7ee952d66b9f3"
+    # BEWARE: get_url + file is not atomic. So if the following lines are
+    # uncommented, and the 'file' part of the task fails, the whole
+    # task fails but on next run it will succeed because the precondition
+    # dest+checksum is already satisfied. Thus, we need to put the
+    # file module on a separate task
+    # owner: root
+    # group: wheel
+    # mode: ug+x,u+s
```
Comment (enricostano) [inline provisioning/roles/container.osx/tasks/xhyve.yml]: Can we remove this comments?

Comment (rewritten) [inline provisioning/roles/container.osx/tasks/xhyve.yml]: Why? It makes sense, as it seems logical to merge the two commands into one, to tell any future developer the exact reason we don't use owner/group/mode in get_url command.

Maybe just link to the issue?

Comment (rewritten) [inline provisioning/roles/container.osx/tasks/xhyve.yml]: https://github.com/ansible/ansible/issues/11821

Comment (enricostano) [inline provisioning/roles/container.osx/tasks/xhyve.yml]: Just leave the comment in next task IMHO

Comment (rewritten) [inline provisioning/roles/container.osx/tasks/xhyve.yml]: Done

File: provisioning/roles/container.osx/tasks/main.yml (hunk 4 above)
Comment (rewritten) [inline provisioning/roles/container.osx/tasks/main.yml]: The only use of ruby is to install homebrew. System ruby is enough for that.

**Comment by rewritten** [comment]
Linux in a container

File: provisioning/roles/container.osx/tasks/main.yml (hunk 4 above)
Comment (enricostano) [inline provisioning/roles/container.osx/tasks/main.yml]: :+1:

**Comment by rewritten** [comment]
I can confirm that I access my files by nfs on the container machine. Now to learn everything.

**Comment by rewritten** [comment]
captura de pantalla 2016-01-25 a las 0 27 47

**Comment by rewritten** [comment]
Maybe merge this?

**Comment by rewritten** [comment]
I'll squash everything when OK

**Comment by markets** [comment]
@rewritten I have 0 knowledge of Ansible :sweat_smile:, so for me :+1: 

Maybe be should deprecate the current Vagrant support?

**Comment by rewritten** [comment]
@markets as soon as we have the linux versions of the container set up yes, vagrant is going away.

**Comment by enricostano** [comment]
:+1:

**Comment by d2bit** [comment]
I thought you were using ansible to set a docker container up to serve the rails app. :confused: 
Installing ansible in my machine is not cool :disappointed: 
I've opened a PR to show you how my environment looks like.. https://github.com/coopdevs/timeoverflow/pull/187

**Comment by rewritten** [comment]
@d2bit It is just the first steps to have ansible set up a docker container to have a "local" development environment. Baby steps. Here we've just got to run a container virtual machine, next we'll use whatever (your compose file maybe) to set up the containers.

**Comment by rewritten** [comment]
@d2bit why not having ansible in your machine? It's just a python package

**Comment by enricostano** [comment]
The main idea is to use Ansible to do the following:
 - setup development environment on local machine (Linux / OS X)
 - provision development environment
 - provision production environment

right @rewritten ?

**Comment by rewritten** [comment]
Yes, and possibly point 2 and 3 will be pretty similar. At least for what respects the dependencies (we may give one container many roles in development, like in katuma, and one container per role in production, to ease scaling out, or we might have everything separated always. Not sure about the implications for speed)

**Comment by enricostano** [comment]
:muscle: :+1:

**Comment by rewritten** [comment]
But right now I'm missing the steps of "creating the container inside the machine", and that's probably what @d2bit 's docker script does. Is it compatible with lxc ?

**Comment by enricostano** [comment]
We can either use LXC or docker as box for dev environment, they are not compatible per se. but it should just be as simple as choosing which provider do you want to use as box. Are you in OS X? Then use the `xhyve` + `docker`. Are you on Linux? You can choose between `LXC` (simpler and already built in Linux kernel) or `docker`.

Once you setup your dev env box, you can provision it and you're ready to go.

**Comment by d2bit** [comment]
@rewritten I don't like installing ansible in my OS, 'cause I don't want to install things I probably won't need soon.
I like to work with containers for services, so I don't have to install postgres, redis, mongodb... on my OS.
_No se si me explico._

**Comment by rewritten** [comment]
@d2bit totally explicado :wink: .

The local development set up is of course totally optional, and it's great that you already have a sane environment. I really hope to be able to use your docker config to help building a lxc/xhyve based set up. 

If anyone prefers a different dev set-up, I would assume they have enough knowledge to set up themselves, and use the ansible recipes as a list of system dependencies, with no need to actually execute them.

On the contrary I will be deploying this until we have a CI that's able to do that alone, so I need ansible anyway, thus it will be the "supported" way to build a local env. Besides, I will keep the hxyve version up-to-date, and Enrico will keep the lxc one, so we are covered.

**Comment by rewritten** [comment]
> I thought you were using ansible to set a docker container up to serve the rails app.
@d2bit it's exactly what we are trying to do.

**Comment by d2bit** [comment]
I'll keep an eye on this, and take a look to how ansible plays to docker provision.
But seems enrico is far beyond my knowledge.
I'll try to be helpful here :muscle:

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -42,4 +42,28 @@ Actualizar rubygems e instalar bundler
...
+`export RUBYGEMS_GEMDEPS="-"` en ningún fichero de inicio:
+- `.bashrc`
+- `.zshrc`
+- `.zprofile`
+- `.profile`
+- `.bash_profile`
+- `.zlogin`
+- `.my_init_var_script`
+
+etc
```
Comment (saimonmoore) [inline CONTRIBUTING.md]: Is there a strong reason why this is in spanish?

**Comment by saimonmoore** [comment]
looks good :+1:

File: CONTRIBUTING.md (hunk 6 above)
Comment (rewritten) [inline CONTRIBUTING.md]: :sweat: no... it's just that until now everyone in the project was based around spanish/iberian countries (including CAT)

Comment (enricostano) [inline CONTRIBUTING.md]: `EN` by default for "code" I would say

**Comment by rewritten** [comment]
@d2bit in the end it is easier to use docker, as native dhyve and similar do not seem mature enough to be scripted. So thank you for your configuration file, it's been the base for this.

@enricostano  http://metal.timeoverflow.org   nginx running inside a container, exposing 80:80

**Comment by rewritten** [comment]
More notes: docker in Linux is setting up containers, not virtual machines, so once the dependencies are clear, we will have almost parity (with boot2docker mounting /Users on Mac)

**Comment by rewritten** [comment]
This concept needs to be revisited, and gain support for the chosen ansible-based provisioning system. I'll close this in favour of a complete rework.

**Pull request closed without merging**